### PR TITLE
 [MM-47326]: Contribute - WYSIWYG editor 

### PR DESCRIPTION
**Pull request by michelengelen**
#### Summary
This will add the WYSIWYG editor to the webapp.

For now this is behind a feature flag and a setting in the advanced section since the editor as it is now is pretty much in an alpha state. In an effort to keep working on this in smaller iterations to be more effective the team decided we can release this as is for now and deal with any bugs as we go on.

to enable it you have to set the feature flag: `MM_FEATUREFLAGS_WYSIWYGEDITOR=true`, then the setting in the advanced settings will appear and you can use it in the create post section

#### Ticket Link
EPIC: [MM-47326](https://mattermost.atlassian.net/browse/MM-47326)

#### Related Pull Requests
- Has server changes: https://github.com/mattermost/mattermost-server/pull/22020

This is a re-opened PR from #11186 

#### Release Note
```release-note
Adds the alpha state of the WYSIWYG editor to the create post section behind feature flag and advanced setting
```

[MM-47326]: https://mattermost.atlassian.net/browse/MM-47326?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@michelengelen this is awesome. A few more issues I found while testing today:

1. The input field's non-focused state doesn't seem to be right. It looks like we lost this state from the input. When not focused, the border should 16% opacity, and only when focused should it be 32%.
2. It doesn't seem to be honoring the setting for command-enter in `Settings > Advanced`. I have this set to 'On for all messages', but when I press enter it sends instead of adding a new line.

image

3. As mentioned previously, the Command-option-K shortcut for inserting a link is not working
4. Still need to implement the upload shortcut (Command-U)
5. The strikethrough formatting option should show next to the italic button
6. The shortcut for strikethrough (Shift-Option-X) doesn't seem to be working
7. The shortcut for code snippet (Command-Option-C) doesn't seem to be working
8. The shortcut for code block is showing in the tooltip as Command-Option-C which appears to be the same as code snippet. I guess we need to choose which one gets this shortcut and provide an alternate for one of them? In the product currently, Command-Option-C is used for the code snippet, so we should find an alternate for code blocks (or remove it for now until we can come up with another option)
9. In the link popover, I'm now seeing two icons that represent ENTER.
    image

10. While inside a code block, pressing down on the last line of code is no longer adding a new line outside the code block. Similarly, pressing up on the first line of code is no longer moving the cursor above the code block.
11. When selecting text on the right side and clicking the link button, the link popover opens too far on the right side and gets cut off
    image
12. In the emoji autocomplete, I'm seeing `SUGGESTION.MENTION.EMOJI` as the title which I assume is a translation ID.
    image
13. When I open the emoji picker (by clicking the emoji icon on the right side), the `ESC` key does not close the picker.
14. Adding attachments to the input field works, but when I post the message, the attachment doesn't end up posting with the message.
15. Similarly, I am not able to post a message that only has an attachment and not text - the send button is disabled

### Non-blocking comments
- Should we consider linking `@mentions` and formatting them as links to be true WYSIWYG
- The send button needs a hover state. It seems like it lost its hover state (or it never had one?)
- In code blocks, the language label isn't quite aligned properly with the chevron
    image
- The size of emojis looks to be slightly different with the WYSIWYG editor vs. what gets rendered in the post
    image


### To be addressed separately
- add WYSIWYG editor to edit post
- add WYSIWYG editor the RHS and threads view reply input
- other places WYSIWYG could be added
- markdown parsing on paste

**Comment by matthewbirtch** [comment]
Looks like you fixed most of the issues @michelengelen. 

A couple more things:
1. For the enter icon, can we adjust the alignment a bit and add a few px of space between the icon and the word ENTER?
image

2. Also on smaller screens, we're not collapsing any formatting buttons by the looks of it like we do with the current advanced text editor. Things seem to be bleeding out of the container. Is this something you'll work on in a separate PR?
image

**Comment by michelengelen** [comment]
@matthewbirtch same procedure as last time:
 
>     1. For the enter icon, can we adjust the alignment a bit and add a few px of space between the icon and the word ENTER?

Fixed!

>     2. Also on smaller screens, we're not collapsing any formatting buttons by the looks of it like we do with the current advanced text editor. Things seem to be bleeding out of the container. Is this something you'll work on in a separate PR?

This is a bit hard to do without excessive calculations. I would like to try and find the best solution possible. I will take this on in the second iteration (same as the hidden formatting bar feature)

On a side note: 
1. I fixed the `ArrowDown` behavior in code blocks. Still missing the `ArrowUp` behavior though, which is a lot harder to do
2. I fixed the emoji sizing in the editor ... should now be the same as in posts (except for emoji posts)

**Comment by esethna** [comment]
@michelengelen really great work, inot sure how many of these have already been found but here's a few things I noticed:

1. When selecting a code block, it defaults to CSS (can we default to no specific language selected but still have the dropdown available for a language selection (may need a placeholder text?)?

2. The language selector might be broken:
image

3. Message priority button doesn't seem to be working

4. When tables are used, can we allow rows & columns in the table, followed by the delete key to delete the selected row/column? I had a hard time figuring out how to remove a column.

5. With tables, can we allow ctrl + A or a dragging across the table to select it for deletion. I found it a little finicky to figure out to properly select the table so it could be deleted

6. I think this was discussed already, but it would be really excellent if the table pop-over could be shown in context with the table when a selection is made in the table. 

7. After using the link button and entering a link, can we move the cursor to after the word (see google docs)

8. Any change we can implement this improvement along with the new editor? ;) https://mattermost.atlassian.net/browse/MM-48061

**Comment by michelengelen** [comment]
@esethna Comments between the lines.

>     1. When selecting a code block, it defaults to CSS (can we default to no specific language selected but still have the dropdown available for a language selection (may need a placeholder text?)?

FIXED
 
>     2. Language selector should validate languages: https://gyazo.com/c04d68ef2a983a44e39ae1a3c92e9796

FIXED

>     3. The language selector needs to be allowed to span outside the input area:

FIXED

>     4. Message priority button doesn't seem to be working

FIXED

>     5. When tables are used, can we allow rows & columns in the table, followed by the delete key to delete the selected row/column? I had a hard time figuring out how to remove a column.

The whole table feature is still just very basic and the proposed changes are particularly hard to implement. For this we would need to render the whole structure differently. We want to provide a similar experience as with the editor in JIRA/Confluence, but as this is such a big task I would like to push this to a later iteration.

>     6. With tables, can we allow ctrl + A or a dragging across the table to select it for deletion. I found it a little finicky to figure out to properly select the table so it could be deleted

see comment on point 5.

>     7. I think this was discussed already, but it would be really excellent if the table pop-over could be shown in context with the table when a selection is made in the table.

see comment on point 5.

>     8. After using the link button and entering a link, can we move the cursor to after the word (see google docs)

FIXED

>     9. Any chance we can implement this improvement along with the new editor? ;) https://mattermost.atlassian.net/browse/MM-48061

This is also very hard to do since the trigger functionality has no way of knowing the current state of the editor(s) which would be needed for it. There are a few ways of doing this, but it is definitely out of scope for this PR

Let me know what you think

**Comment by matthewbirtch** [comment]
Seems like there might be something wrong with the menu for code block language selection. The height is very short and can't see any items within it:
image

**Comment by michelengelen** [comment]
@matthewbirtch this should be fixed by now ... the test server did not update and that's why it is currently broken

**Comment by michelengelen** [comment]
@matthewbirtch @esethna the test server is updated with the newest commit

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Seeing few issues 
1. Attachments remain in textbox after posting
Screen Shot 2023-01-17 at 12 36 24 PM

2. Emoji picker is displaying off screen in Safari. Hidden when RHS is open
Screen Shot 2023-01-17 at 2 54 04 PM

3. Safari: Cursor missing when options other than Normal Text are toggled 
4. Existing posts appear to shift down when typing or deleting text


5. Code language selector is too tall, top items are out of view
Screen Shot 2023-01-17 at 3 15 19 PM

6. When pasting code block, it is not recognized and formatted ( this may be done later). I'm unable to get out of code formatting and add normal text after pasting something similar to screenshot, can't start a new line
Screen Shot 2023-01-17 at 3 25 38 PM

7. Horizontal cursor when table is present

Screen Shot 2023-01-17 at 3 33 39 PM

8. Table input cursor jumps from top cells to bottom right when user stops typing

https://user-images.githubusercontent.com/52937121/213007669-079157b5-c833-4469-a93f-4e9978f3d700.mov

9. Unable to turn linking off e.g. add link and hit enter, add space and try to toggle link off to add more text
10.  Safari: Unable to add link, hitting Enter while on the link popover doesn't do anything 
11. There is slight downward shift to main channel posts (existing) when opening post priority menu. I have seen something similar with some other actions/posts. Maybe the same as issue 4.

https://user-images.githubusercontent.com/52937121/213014071-91864432-7c19-49d3-b404-0a858e08e950.mov

**Comment by michelengelen** [comment]
Hi @jgilliam17 ... thanks for going over this. Comments between the lines as usual:

>     1. Attachments remain in textbox after posting

FIXED

>     2. Emoji picker is displaying off screen in Safari. Hidden when RHS is open

FIXED

>     3. Safari: Cursor missing when options other than Normal Text are toggled

not sure what is causing this ... did not manage to fix for this iteration. @jgilliam17 could you open a ticket in the epic and assign this to me so I can work on this in the second iteration? Thanks!

>     4. Existing posts appear to shift down when typing or deleting text

FIXED

>     5. Code language selector is too tall, top items are out of view

FIXED (applied the Suggestion lists max-height value here - @matthewbirtch is that OK for you?)

>     6. When pasting code block, it is not recognized and formatted ( this may be done later). I'm unable to get out of code formatting and add normal text after pasting something similar to screenshot, can't start a new line

formatting pasted text (at least the more complex stuff) will be handled in one of the next iterations.

>     7. Horizontal cursor when table is present

This is the gapcursor and allows for easy deletion of the table when hitting Backspace

>     8. Table input cursor jumps from top cells to bottom right when user stops typing

I could not reproduce this (tested in FF, FF Dev Edition, Chrome and Safari)

>     9. Unable to turn linking off e.g. add link and hit enter, add space and try to toggle link off to add more text

You can exit the link with the Right-Arrow. This is the default behavior and since chars like dots and commas are valid chars for a link it is hard to determine when we want to exit the link state. For spaces: This is valid for text that has a link applied, so when you have a pure URL like `http://www.mattermost.com` it can be exited with a space.

>     10. Safari: Unable to add link, hitting Enter while on the link popover doesn't do anything

FIXED

>     11. There is slight downward shift to main channel posts (existing) when opening post priority menu. I have seen something similar with some other actions/posts. Maybe the same as issue 4.

FIXED (same issue as no.4)

> Edited: 12. Console error and banner. I am not 100% sure of the steps, but I think I had some text highlighted or pasted, follow by selecting to add table (happened on Desktop daily build)

Could not reproduce ... If you can reproduce and know the steps please open up a ticket.

Thanks for the review... Ready for another round! 🚀

**Comment by matthewbirtch** [comment]
> > ```
> > 5. Code language selector is too tall, top items are out of view
> > ```
> 
> 
> FIXED (applied the Suggestion lists max-height value here - @matthewbirtch is that OK for you?)

Yep, this is good now

**Comment by matthewbirtch** [comment]
> > ```
> > 7. Horizontal cursor when table is present
> > ```
> 
> 
> This is the gapcursor and allows for easy deletion of the table when hitting Backspace

@michelengelen I don't know if this is a common behavior - it was unexpected for me as well. I would have expected it just to be a normal text cursor on a new line (like we do for code blocks)

**Comment by michelengelen** [comment]
> > > ```
> 
> > > 7. Horizontal cursor when table is present
> 
> > > ```
> 
> > 
> 
> > 
> 
> > This is the gapcursor and allows for easy deletion of the table when hitting Backspace
> 
> 
> 
> @michelengelen I don't know if this is a common behavior - it was unexpected for me as well. I would have expected it just to be a normal text cursor on a new line (like we do for code blocks)

I have seen this in a few other wysiwyg editors. I will Check if a normal caret is possible as well. This is the standard behavior for the package we are using.

**Comment by jgilliam17** [comment]
Thanks @michelengelen for the quick fix
Looks good, for alpha release to community
Noticed one small issue with Upload limit error, missing red font color on main channel, not reproducing on RHS.
Screen Shot 2023-01-19 at 1 23 18 PM

**Comment by michelengelen** [comment]
@jgilliam17 I fixed the error text color and added the `[user] is typing...` element as well.

@matthewbirtch the gapcursor is currently needed since when removing it from the extensions we cannot exit a table anymore when there is no text around it. I will leave it in for now, so we can see if this is acceptable in daily usage, or find another solution.

Thanks for your reviews!

Now its time for the devs to rip apart my code! 🤔

@mickmister @koox00 @AshishDhama and @Willyfrog: There are a few functions inside the `advanced_create_post.tsx` file that I have recreated to fit with the new data-flow. some just extract the message value from the event, where others are getting the string from the wysiwyg and need to loop that through (similar to the "old" implementations).

HAVE FUN!!!

**Comment by matthewbirtch** [comment]
> added the `[user] is typing...` element as well.

The user is typing element is too small, but I believe that was going to be fixed in another PR since we identified that elsewhere, right?

**Review by matthewbirtch — APPROVED** [review]
Approving this for alpha release (on community) knowing there are polish and follow-up tickets/PRs to come. Awesome work @michelengelen

A few items to add to your list of follow-ups:
- I noticed that in the editor I can add a code block within a table cell. We should disable code blocks within tables since they won't properly render within the cell when posted anyway
- Same goes for bulleted lists, numbered lists, and quotes. They don't seem to render properly when added to table cells

**Comment by jgilliam17** [comment]
@michelengelen  additional note on `[user] is typing...` element is not visible for me on main channel. It is visible on RHS

**Comment by AshishDhama** [comment]
@michelengele multiple checks failing, could please update the pr

**Comment by michelengelen** [comment]
@jgilliam17 strange ... its the same component ... could you try retesting it?

File: package.json [hunk 1]
```diff
@@ -88,6 +136,10 @@
     "redux-persist": "6.0.0",
     "redux-thunk": "2.4.2",
     "regenerator-runtime": "0.13.10",
+    "rehype-stringify": "9.0.3",
```
Comment (hmhealey) [inline package.json:139]: This is an awful lot of extra dependencies for just generating Markdown. I'm going to need to look through those in some more detail, but are you sure they're all needed?

File: package.json [hunk 2]
```diff
@@ -20,6 +20,39 @@
     "@stripe/react-stripe-js": "1.13.0",
     "@stripe/stripe-js": "1.41.0",
     "@tippyjs/react": "4.2.6",
+    "@tiptap/core": "2.0.0-beta.209",
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply package.json:23]: Could we perhaps use Tiptap's StarterKit package to reduce the number of extensions here? https://tiptap.dev/api/extensions/starter-kit

File: packages/mattermost-redux/src/selectors/entities/preferences.ts [hunk 3]
```diff
@@ -268,6 +268,25 @@ export function getHasDismissedSystemConsoleLimitReached(state: GlobalState): bo
     return getBool(state, Preferences.CATEGORY_UPGRADE_CLOUD, Preferences.SYSTEM_CONSOLE_LIMIT_REACHED, false);
 }
 
+export function getWysiwygPreference(state: GlobalState): 'true' | 'false' {
+    return get(
```
Comment (hmhealey) [inline packages/mattermost-redux/src/selectors/entities/preferences.ts:272]: This looks like it could use `getBool` instead of returning a string

File: packages/types/tsconfig.json [hunk 4]
```diff
@@ -14,5 +14,6 @@
         "rootDir": "./src",
         "composite": true
     },
-    "exclude": ["**/node_modules", "**/lib", "**/*.test.js", "**/*.test.ts"]
+    "exclude": ["**/node_modules", "**/lib", "**/*.test.js", "**/*.test.ts"],
+    "typeRoots": ["node_modules/@types"],
```
Comment (hmhealey) [inline packages/types/tsconfig.json:18]: What is this needed for?

File: reducers/drafts.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    case UserTypes.LOGOUT_SUCCESS:
+        return {};
+    default:
+        return state;
+    }
+}
+
+export default combineReducers({
+    byChannel: persistReducer({key: 'drafts.byChannel', storage: localForage}, byChannel),
+    byThread: persistReducer({key: 'drafts.byThread', storage: localForage}, byThread),
```
Comment (hmhealey) [inline reducers/drafts.ts:77]: Awesome, I've wanted to do this for ages. Having drafts stored in `state.storage` has actually caused quite a few performance issues in the past, so getting them stored in their own section of the state is awesome.

Does this work with the global drafts view and server-side syncing of drafts? Does it also support migrating old drafts to the new location/format on upgrade?

File: sass/components/_channel-switcher.scss [hunk 6]
```diff
@@ -50,7 +50,9 @@
 
     .channel-switcher__suggestion-box {
         position: relative;
-        height: 362px;
+        height: 100%;
```
Comment (hmhealey) [inline sass/components/_channel-switcher.scss:53]: Is this change supposed to be part of this PR?

File: tsconfig.json [hunk 7]
```diff
@@ -23,6 +23,7 @@
     "baseUrl": ".",
     "useUnknownInCatchVariables": false,
     "paths": {
+        "@tiptap/core": ["node_modules/@tiptap/core/src/index"],
```
Comment (hmhealey) [inline tsconfig.json:26]: Is this needed? I wouldn't expect this sort of import rewriting to be needed for this library. The fact that we do it for the other packages is something specific to the monorepo

File: utils/constants.tsx [hunk 8]
```diff
@@ -2209,6 +2214,13 @@ export const InsightsCardTitles = {
     },
 };
 
+export enum WysiwygPluginNames {
```
Comment (hmhealey) [inline utils/constants.tsx:2217]: Do these need to live here or is there somewhere they could closer to the WYSIWYG editor itself? It looks like they're only used within that folder, so I'd like to see them moved there

File: webpack.config.js [hunk 9]
```diff
@@ -139,6 +139,12 @@ var config = {
                     },
                 ],
             },
+            {
+                test: /\.m?js/,
```
Comment (hmhealey) [inline webpack.config.js:143]: What is this change for? There's a few things I notice that are potentially incorrect here:
1. This can capture plain old `.js` files because the `m` is optional which seems like we wouldn't want to have happen
2. Similarly, it can also capture `.jsx` files because it doesn't check for the end of the string with `$`

File: components/emoji_picker/components/emoji_picker_kaomoji.tsx [hunk 10]
```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {ChangeEvent, KeyboardEvent, useEffect, useRef, useState} from 'react';
```
Comment (hmhealey) [inline components/emoji_picker/components/emoji_picker_kaomoji.tsx:4]: Could you pull this out and submit it as a separate PR? I'm 0/5 on if we add it, but this PR's already massive, so we should try to pare it down wherever possible

**Review by hmhealey — CHANGES_REQUESTED** [review]
I haven't had a chance to look through the components themselves yet with this, but here's a review of the more infrastructure parts of it at least

**Comment by jgilliam17** [comment]
> @jgilliam17 strange ... its the same component ... could you try retesting it?

I checked today again and still miss the `user typing` on main channel

File: package.json [hunk 11]
```diff
@@ -51,16 +86,29 @@
...
+    "mdast-util-gfm": "2.0.1",
+    "mdast-util-to-markdown": "1.4.0",
     "memoize-one": "6.0.0",
     "moment-timezone": "0.5.38",
     "p-queue": "7.3.0",
     "pdfjs-dist": "2.15.349",
     "popper.js": "1.16.1",
     "process": "0.11.10",
     "prop-types": "15.8.1",
+    "prosemirror-dropcursor": "1.6.1",
```
[the remark this answers is not in the document]
Comment (michelengelen) [reply package.json:102]: For some reason the peerDependencies do not work/are not installed when not adding them manually

File: package.json (hunk 1 above)
Comment (michelengelen) [reply package.json:139]: the problem is that we need to go through a few steps to convert the data in the editor to MD.

HTML => HAST => MDAST => MD 

and the other way around. Although I am currently not using that functionality

File: packages/types/tsconfig.json (hunk 4 above)
Comment (michelengelen) [reply packages/types/tsconfig.json:18]: good question ... thats none of my changes ... might be a broken merge

File: reducers/drafts.ts (hunk 5 above)
Comment (michelengelen) [reply reducers/drafts.ts:77]: thanks for the compliment, but this as well is not my change ... it has to be a broken merge here

File: sass/components/_channel-switcher.scss (hunk 6 above)
Comment (michelengelen) [reply sass/components/_channel-switcher.scss:53]: same here

File: tsconfig.json (hunk 7 above)
Comment (michelengelen) [reply tsconfig.json:26]: it was needed, but apparently they fixed it with the new version

File: webpack.config.js (hunk 9 above)
Comment (michelengelen) [reply webpack.config.js:143]: fixed

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (michelengelen) [reply package.json:23]: we could, but there is a ton of stuff in there we don't really need ... I kind of copied their approach and added everything they did add, but wanted to try and remove the stuff we don't need later on

File: package.json (hunk 1 above)
Comment (michelengelen) [reply package.json:141]: fixed

File: tsconfig.json (hunk 7 above)
Comment (michelengelen) [reply tsconfig.json:26]: Nope, they did not ... the types get not resolved correctly and therefor the `check-types` pipeline is failing

**Comment by michelengelen** [comment]
@hmhealey I have fixed the tests, removed the packages causing the problems, added a custom parser logic and included some tests for it as well. I did remove the Kaomoji-Stuff from this PR as well and will add a new PR in the following days!

@AshishDhama @mickmister @Willyfrog and @koox00 all tests are green now, so feel free to start reviewing

@jgilliam17 I will fix that `[user] is typing ...` issue next week.

Thanks all! :shipit:

File: types/store/draft.ts [hunk 12]
```diff
@@ -22,10 +24,10 @@ export type PostDraft = {
...
-            priority: PostPriority|'';
-            requested_ack?: boolean;
-            persistent_notifications?: boolean;
-        };
+        priority?: PostPriorityMetadata;
     };
 };
+
+export type NewPostDraft = PostDraft & {
+    content?: JSONContent;
```
Comment (AshishDhama) [inline types/store/draft.ts:32]: @michelengelen why do we need to introduce a new type if you add an optional key? I am not sure we are doing version bump for types, but we should update the existing `PostDraft.`
but would leave that decision with web platform cc// @hmhealey

File: tests/setup.js [hunk 13]
```diff
@@ -8,7 +8,7 @@ import '@testing-library/jest-dom';
 
 import './redux-persist_mock';
 import './react-intl_mock';
-import './react-router-dom_mock';
+import './react-router-dom_mock.ts';
```
Comment (AshishDhama) [inline tests/setup.js:11]: I don’t think this is required If I am not wrong we by default search for ts file.

Comment (michelengelen) [reply tests/setup.js:11]: it is needed, since importing `.ts` in `.js` files and vice versa need the extension

File: types/store/draft.ts (hunk 12 above)
Comment (michelengelen) [reply types/store/draft.ts:32]: good question ... this is a change that came from the working group during mattercon, so I am not sure who introduced that ... I can just change the existing type, since the comment is valid! :D
